source/extensions/filters/common/ratelimit/ratelimit_impl.cc

```diff
@@ -0,0 +1,148 @@
+  ASSERT(callbacks_ != nullptr);
+  request_->cancel();
+  callbacks_ = nullptr;
+}
+
+void GrpcTlsClientImpl::createRequest(
+    envoy::service::ratelimit::v2::RateLimitRequest& request, const std::string& domain,
+    const std::vector<Envoy::RateLimit::Descriptor>& descriptors) {
+  request.set_domain(domain);
+  for (const Envoy::RateLimit::Descriptor& descriptor : descriptors) {
+    envoy::api::v2::ratelimit::RateLimitDescriptor* new_descriptor = request.add_descriptors();
+    for (const Envoy::RateLimit::DescriptorEntry& entry : descriptor.entries_) {
+      envoy::api::v2::ratelimit::RateLimitDescriptor::Entry* new_entry =
+          new_descriptor->add_entries();
+      new_entry->set_key(entry.key_);
+      new_entry->set_value(entry.value_);
+    }
+  }
+}
+
+void GrpcTlsClientImpl::limit(RequestCallbacks& callbacks, const std::string& domain,
+                              const std::vector<Envoy::RateLimit::Descriptor>& descriptors,
+                              Tracing::Span& parent_span) {
+  ASSERT(callbacks_ == nullptr);
+  callbacks_ = &callbacks;
+
+  envoy::service::ratelimit::v2::RateLimitRequest request;
+  createRequest(request, domain, descriptors);
+  request_ =
+      async_client_->send(service_method_, request, *this, parent_span, shared_state_->timeout_ms_);
+}
+
+void GrpcTlsClientImpl::onSuccess(
+    std::unique_ptr<envoy::service::ratelimit::v2::RateLimitResponse>&& response,
+    Tracing::Span& span) {
+  LimitStatus status = LimitStatus::OK;
+  ASSERT(response->overall_code() != envoy::service::ratelimit::v2::RateLimitResponse_Code_UNKNOWN);
+  if (response->overall_code() ==
+      envoy::service::ratelimit::v2::RateLimitResponse_Code_OVER_LIMIT) {
+    status = LimitStatus::OverLimit;
+    span.setTag(Constants::get().TraceStatus, Constants::get().TraceOverLimit);
+  } else {
+    span.setTag(Constants::get().TraceStatus, Constants::get().TraceOk);
+  }
+
+  Http::HeaderMapPtr headers = std::make_unique<Http::HeaderMapImpl>();
+  if (response->headers_size()) {
+    for (const auto& h : response->headers()) {
+      headers->addCopy(Http::LowerCaseString(h.key()), h.value());
+    }
+  }
+  callbacks_->complete(status, std::move(headers));
+  callbacks_ = nullptr;
+}
+
+void GrpcTlsClientImpl::onFailure(Grpc::Status::GrpcStatus status, const std::string&,
+                                  Tracing::Span&) {
+  ASSERT(status != Grpc::Status::GrpcStatus::Ok);
+  callbacks_->complete(LimitStatus::Error, nullptr);
+  callbacks_ = nullptr;
+}
+
+GrpcFactoryImpl::GrpcFactoryImpl(const envoy::config::ratelimit::v2::RateLimitServiceConfig& config,
+                                 Grpc::AsyncClientManager& async_client_manager,
+                                 ThreadLocal::SlotAllocator& tls, Stats::Scope& scope)
+    : tls_(tls) {
+  envoy::api::v2::core::GrpcService grpc_service;
+  grpc_service.MergeFrom(config.grpc_service());
+  // TODO(htuch): cluster_name is deprecated, remove after 1.6.0.
```

Comment:
Not just the code but api also needs to be removed. Not sure if there is an impact on tests.  ithttps://github.com/envoyproxy/envoy/blob/b6ff832b9159fad4aef3e93d6243753f9427494c/api/envoy/config/ratelimit/v2/rls.proto#L23
Probably separate PR so that it is clean? I can do that once this is done.